src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,123 @@
+  : _probes(max_api_key() + 2)
+  , _unknown_handler_index(max_api_key() + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka:handler"),
+      {
+        sm::make_counter(
+          "requests_completed",
+          [this] { return _requests_completed; },
+          sm::description("Number of requests completed"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_errored",
+          [this] { return _requests_errored; },
+          sm::description("Number of requests errored"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_in_progress",
+          [this] { return _requests_in_progress_every_ns / 1'000'000'000; },
+          sm::description("Number of requests in progress"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_received",
+          [this] { return _bytes_received; },
+          sm::description("Number of bytes received from requests"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_sent",
+          [this] { return _bytes_sent; },
+          sm::description("Number of bytes sent in replies"),
+          labels)
+          .aggregate(aggregate_labels),
+      });
+}
+
+/*
+ * This roughly approximates an integral of `_requests_in_progress`.
+ * By providing Prometheus with a counter of the integral rather than
+ * a gauge for `_requests_in_progress` we avoid any bias in the value
+ * that Prometheus's sampling rate could introduce.
+ */
+void handler_probe::sample_in_progress() {
+    auto now = ss::lowres_clock::now();
+    auto s_diff = std::chrono::duration_cast<std::chrono::nanoseconds>(
```

Comment:
Note that you don't need duration_cast here or in many places you see it: it's only need to coerce durations where there is some loss of precision or something like that. In this case just constructing nanoseconds directly from the underlying duration will work.

Or even better, short and sweet: `(now - _last_recorded_in_progress) / 1ns` can replace the whole thing

Reply:
Replaced it with the suggested line.